katas/content/teleportation/Common.qs

```diff
@@ -1,7 +1,123 @@
+    operation ComposeTeleportation(
+        bellPrepOp : ((Qubit, Qubit) => Unit), 
+        getDescriptionOp : ((Qubit, Qubit) => (Bool, Bool)), 
+        reconstructOp : ((Qubit, (Bool, Bool)) => Unit),
+        qAlice : Qubit, 
+        qBob : Qubit, 
+        qMessage : Qubit) : Unit {
+
+        bellPrepOp(qAlice, qBob);
+        let classicalBits = getDescriptionOp(qAlice, qMessage);
+        
+        // Alice sends the classical bits to Bob.
+        // Bob uses these bits to transform his part of the entangled pair into the message.
+        reconstructOp(qBob, classicalBits);
+    }
+    
+    operation SendMessage_Reference(qAlice: Qubit, qMessage: Qubit) : (Bool, Bool) {
+        CNOT(qMessage, qAlice);
+        H(qMessage);
+        return (M(qMessage) == One, M(qAlice) == One);
+    }
+
+    operation ReconstructMessage_Reference(qBob : Qubit, (b1 : Bool, b2 : Bool)) : Unit {
+        if b1 {
+            Z(qBob);
+        }
+        if b2 {
+            X(qBob);
+        }
+    }
+
+    // ------------------------------------------------------
+    // Helper operation that runs a teleportation operation (specified by teleportOp).
+    // The state to teleport is set up using an operation (specified by setupPsiOp).
+    //
+    // Specifying the state to teleport through an operation allows to get the inverse
+    // which makes testing easier.
+    operation TeleportTestHelper(
+        teleportOp : ((Qubit, Qubit, Qubit) => Unit), 
+        setupPsiOp : (Qubit => Unit is Adj)) : Bool {
+        
+        use (qMessage, qAlice, qBob) = (Qubit(), Qubit(), Qubit());
+        setupPsiOp(qMessage);
+            
+        // This should modify qBob to be identical to the state
+        // of qMessage before the function call.
+        teleportOp(qAlice, qBob, qMessage);
+            
+        // Applying the inverse of the setup operation to qBob
+        // should make it Zero.
+        Adjoint setupPsiOp(qBob);
+        if not CheckZero(qBob){
+            Message("Incorrect. The state was teleported incorrectly.");
+            DumpMachine();
+            ResetAll([qMessage, qAlice, qBob]);
+            return false;
+        }
+        ResetAll([qMessage, qAlice, qBob]);
+        Message("Correct.");
+        return true;
+    }
+
+    // ------------------------------------------------------
+    // Run teleportation for a number of different states.
+    // After each teleportation success is asserted.
+    // Also repeats for each state several times as
+    // code is expected to take different paths each time because
+    // measurements done by Alice are not deterministic.
+    operation TeleportTestLoop(teleportOp : ((Qubit, Qubit, Qubit) => Unit)) : Bool {
+        // Define setup operations for the message qubit
+        // on which to test teleportation: |0⟩, |1⟩, |0⟩ + |1⟩, unequal superposition.
+        let setupPsiOps = [I, X, H, Ry(ArcCos(0.6) * 2.0, _)];
```

Comment:
This code is much easier to follow now! Let's bring back the string names of the states to help troubleshoot; we can do this by making this array an array of tuples
```suggestion
        let setupPsiOps = [(I, "|0⟩"), (X, "|1⟩"), (H, "|+⟩"), (Ry(ArcCos(0.6) * 2.0, _), "0.6|0⟩ + 0.8|1⟩")];
```
and then changing the first for loop to for (psiOp, psiName) and passing an extra string psiName to TeleportTestHelper that is then printed in the failure message